File: src/agentswarm/context.py
"""The per-run context shared by the primitives.

A :class:`RunContext` is the single object a running workflow talks to,
indirectly, through the primitives. It bundles the wired-up layers (bridge,
scheduler), the control and event sink, the run's ``args``, and the mutable
display state (``meta``, ``current_phase``).

:func:`build_context` is the one place that wires the layers together from a
:class:`Config`, so the runtime worker (and tests) get a ready-to-use context
in a single call.
"""

from __future__ import annotations

import os
from dataclasses import dataclass, field
from typing import Any

from .adapters import Config
from .bridge import Bridge
from .control import Control, NullControl
from .events import EventSink, NullSink
from .scheduler import Scheduler


@dataclass
class RunContext:
    config: Config
    bridge: Bridge
    scheduler: Scheduler
    control: Control
    sink: EventSink
    args: Any = None
    meta: dict[str, Any] = field(default_factory=dict)
    current_phase: str | None = None

    def emit(self, ev: dict[str, Any]) -> None:
        self.sink.emit(ev)


def build_context(
    config: Config,
    *,
    source: str | os.PathLike[str] | None = None,
    args: Any = None,
    sink: EventSink | None = None,
    control: Control | None = None,
) -> RunContext:
    """Wire a full run context from a config and the run's surroundings."""
    sink = sink or NullSink()
    control = control or NullControl()
    bridge = Bridge(config, source=source)
    scheduler = Scheduler(
        config.settings.resolved_concurrency(),
        config.settings.max_agents,
        checkpoint=control.checkpoint,
    )
    return RunContext(
        config=config,
        bridge=bridge,
        scheduler=scheduler,
        control=control,
        sink=sink,
        args=args,
    )

File: src/context.ts
/**
 * The per-run context shared by the primitives.
 *
 * A {@link RunContext} is the single object a running workflow talks to,
 * indirectly, through the primitives. It bundles the wired-up layers (bridge,
 * scheduler), the control and event sink, the run's `args` and budget target,
 * and the mutable display state (`currentPhase`). {@link buildContext} is the one
 * place that wires the layers together from a {@link Config}.
 */

import { resolveConcurrency } from "./adapters/config.js";
import type { Config } from "./adapters/types.js";
import { Bridge } from "./bridge.js";
import { NullControl, type Control } from "./control.js";
import { BudgetExhausted } from "./errors.js";
import { NullSink, type EventSink, type WorkflowEvent } from "./events.js";
import { Scheduler } from "./scheduler.js";

export interface RunContext {
  config: Config;
  bridge: Bridge;
  scheduler: Scheduler;
  control: Control;
  sink: EventSink;
  args: unknown;
  /** The run's working directory; anchors nested workflow() name resolution. */
  source: string;
  /** The run's token target (for `budget`), or null when none was set. */
  budgetTotal: number | null;
  /**
   * Shared usage tally behind `budget.spent()`. Tokens are ESTIMATED from agent
   * reply text (chars/4) — adapters do not report real usage uniformly — so the
   * budget is an honest approximation, shared across nested workflow() calls.
   */
  usage: { outputChars: number };
  currentPhase: string | null;
  emit(ev: WorkflowEvent): void;
}

/** Estimated tokens spent so far (chars/4 of every agent reply in this run). */
export function spentTokens(usage: { outputChars: number }): number {
  return Math.ceil(usage.outputChars / 4);
}

export interface BuildContextOptions {
  source?: string;
  args?: unknown;
  sink?: EventSink;
  control?: Control;
  budgetTotal?: number | null;
}

/** Wire a full run context from a config and the run's surroundings. */
export function buildContext(config: Config, options: BuildContextOptions = {}): RunContext {
  const sink = options.sink ?? new NullSink();
  const control = options.control ?? new NullControl();
  const bridge = new Bridge(config, { source: options.source });
  const budgetTotal = options.budgetTotal ?? null;
  // The shared tally `budget.spent()` reads. Estimated (chars/4 of agent
  // replies); the guard makes --budget a real ceiling, not advisory.
  const usage = { outputChars: 0 };
  const scheduler = new Scheduler({
    concurrency: resolveConcurrency(config.settings.concurrency),
    maxAgents: config.settings.maxAgents,
    checkpoint: () => control.checkpoint(),
    budgetGuard: () => {
      if (budgetTotal !== null && spentTokens(usage) >= budgetTotal) {
        throw new BudgetExhausted(
          `run reached its budget ceiling of ${budgetTotal} estimated tokens (spent ~${spentTokens(usage)})`,
        );
      }
    },
  });
  return {
    config,
    bridge,
    scheduler,
    control,
    sink,
    args: options.args ?? null,
    source: options.source ?? process.cwd(),
    budgetTotal,
    usage,
    currentPhase: null,
    emit(ev: WorkflowEvent): void {
      sink.emit(ev);
    },
  };
}
